### PR TITLE
HDDS-6891. Add CapacityVolumeChoosingPolicy

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

Currently, the volumeChoosingPolicy is RoundRobin logic, this ticket is to add a new policy that will honor the capacity of the volumes while choosing volumes.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6891

## How was this patch tested?

unit test.

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        .collect(Collectors.toList());
+    if (filtered.size() < 1) {
+      throw new DiskOutOfSpaceException("Out of space: "
+          + "All volumes are less than the container size (=" + maxContainerSize
+          + " B).");
+    } else if (filtered.size() == 1) {
+      return filtered.get(0);
+    } else {
+      int firstIndex = random.nextInt(filtered.size());
+      int secondIndex = random.nextInt(filtered.size());
```
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java:70]: With a small number of volumes, there is a reasonable chance of selecting the same index twice here I think? We might need to do something to ensure you get two unique indexes.

Also what about the case with only 2 volumes available? Do we always want to select the smallest one, or do we want to select the smaller one more often than the other? Maybe we need another if branch for 2 volumes available to handle it specially?

**Review by sodonnel — COMMENTED** [review]
I like this approach because it is so simple. In the HDFS AvailabeSpace policy, there is more complex logic to select the under-used disks, but I feel this approach may work well.

I think I have seen this approach used in other parts of HDFS, perhaps in the namenode when it is choosing between DBs to place a block.

I think we may need to add a bit more logic for 2 available volumes and check we are not picking the same volume twice.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java (hunk 1 above)
Comment (symious) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java:70]: @sodonnel Thanks for the review.

Say we have volume1(available: 200G) and volume2 (available: 100G), the expected result is we use volume1 more often than volume2.

The choosing results could be as follows:
| first choice| second choice | ratio | result |
| ----------- | ---------------- | ----- | ------ |
| volume1   | volume2           | 25% | volume1|
| volume2   | volume1           | 25% | volume1|
| volume1   | volume1           | 25% | volume1|
| volume2   | volume2           | 25% | volume2|

So the ratio of choosing volume1 is 75% and volume2 is 25%. 
I think do not ensure two unique indexes just solves the special case of 2 volumes.

Comment (symious) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/CapacityVolumeChoosingPolicy.java:81]: Sure, added the comment, please have a check.

**Review by sodonnel — APPROVED** [review]
LGTM. Lets give this a try and see how well it works in practice.

**Comment by ferhui** [comment]
@symious Thanks for your contribution. @sodonnel Thanks for your review! Merged

**Pull request merged**